processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java

```diff
@@ -760,6 +872,199 @@ public void close() throws IOException
     }
   }
 
+  public static class CursorOffsetHolder
+  {
+    Offset currOffset = null;
+
+    public Offset get()
+    {
+      return currOffset;
+    }
+
+    public void set(Offset currOffset)
+    {
+      this.currOffset = currOffset;
+    }
+  }
+
+  private static boolean isComparableNullOrEmpty(final Comparable value)
+  {
+    if (value instanceof String) {
+      return Strings.isNullOrEmpty((String) value);
+    }
+    return value == null;
+  }
+
+  private static class CursorOffsetHolderValueMatcherFactory implements ValueMatcherFactory
+  {
+    private final CursorOffsetHolder holder;
+    private final ColumnSelectorFactory cursor;
+
+    public CursorOffsetHolderValueMatcherFactory(
+        CursorOffsetHolder holder,
+        ColumnSelectorFactory cursor
+    )
+    {
+      this.holder = holder;
+      this.cursor = cursor;
+    }
+
+    @Override
+    public ValueMatcher makeValueMatcher(String dimension, final Comparable value)
+    {
+      final DimensionSelector selector = cursor.makeDimensionSelector(
+          new DefaultDimensionSpec(dimension, dimension)
+      );
+
+      final boolean matchNull = isComparableNullOrEmpty(value);
+
+      final int id = selector.lookupId((String) value);
+      if (id < 0) {
+        return new BooleanValueMatcher(false);
```

Comment:
I think the existing code is correct:

null/empty str would still be in the dictionary if there were any rows that had such values, id < 0 for a null/empty value indicates there were rows with value missing/null values

the "matchNull" boolean is there more for catching rows with size 0 when the valueMatcher matches on null.